src/Bicep.Types.Az/Bicep.Types.Az.csproj

```diff
@@ -7,12 +7,36 @@
     <RootNamespace>Azure.Bicep.Types.Az</RootNamespace>
   </PropertyGroup>
 
-  <ItemGroup>
-    <EmbeddedResource Include="../../generated/**/*.json" LogicalName="$([System.String]::new('%(RecursiveDir)%(Filename)%(Extension)').Replace('\', '/').ToLowerInvariant())" />
-  </ItemGroup>
-
   <ItemGroup>
     <ProjectReference Include="../Bicep.Types/Bicep.Types.csproj" />
   </ItemGroup>
 
+  <UsingTask TaskName="ZipDirectory" TaskFactory="RoslynCodeTaskFactory" AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll">
+    <ParameterGroup>
+      <SourceDirectory ParameterType="System.String" Required="true" />
+      <TargetZipFile ParameterType="System.String" Required="true" />
+    </ParameterGroup>
+    <Task>
+      <Using Namespace="System.IO" />
+      <Using Namespace="System.IO.Compression" />
+      <Code Type="Fragment" Language="cs">
+        <![CDATA[
+          Directory.CreateDirectory(Path.GetDirectoryName(TargetZipFile));
+          File.Delete(TargetZipFile);
+          ZipFile.CreateFromDirectory(SourceDirectory, TargetZipFile, CompressionLevel.Optimal, false);
+        ]]>
+      </Code>
+    </Task>
+  </UsingTask>
+
+  <Target Name="EmbedTypes" BeforeTargets="PrepareForBuild">
+    <ItemGroup>
+      <TypeFiles Include="../../generated/**/*.json" />
+    </ItemGroup>
+    <Copy SourceFiles="@(TypeFiles)" DestinationFiles="@(TypeFiles->'$(OutDir)/generated_types/%(RecursiveDir)%(Filename)%(Extension)')" />
+    <ZipDirectory SourceDirectory="$(OutDir)/generated_types" TargetZipFile="$(OutDir)/generated_types.zip" />
```

Comment:
>ZipDirectory [](start = 5, length = 12)

This is going to copy the files and zip them up every time you run the build. Can you split out the copy part and the zip part into separate targets and then declare inputs and outputs on each? This way msbuild will be able to skip the targets if the work has already been done.

Reply:
Declaring Inputs/Outputs on the target worked a charm! I didn't split the target into multiple; what's the benefit of doing that?

Reply:
Nice!

It would be so you can declare inputs and outputs for each stage of processing. In some cases, it can makes you more resilient to things touching your intermediate files without touching the original input files. I guess not a whole lot of benefit in your case, though.